### PR TITLE
Add the flops profiler which measures the time, number of estimated flops and parameters of each module in a model

### PR DESCRIPTION
**Pull request by cli99**

### CODE REVIEW

**Review by jeffra — COMMENTED** [review]
I see DeepSpeedExamples (DSE) has lots of changes here, is this because we're wanting to point to an updated version of DSE that uses the profiler? or just pointing to a different commit than what the current master branch points to?

File: deepspeed/profiling/flops_profiler/examples/bert.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
+from functools import partial
```
Comment (jeffra) [inline deepspeed/profiling/flops_profiler/examples/bert.py:1]: I wonder if we should move these examples to DSE? or maybe still in the DeepSpeed repo but not in the deepspeed package? I am not sure we want them to be installed with the library itself (e.g., hidden away as a library like 
 /usr/local/lib/python3.6/dist-packages/deepspeed) where they are hard to find and run directly.

File: deepspeed/profiling/flops_profiler/profiler.py [hunk 2]
```diff
@@ -0,0 +1,751 @@
...
+
+# BatchNorms
+F.batch_norm = wrapFunc(F.batch_norm, batch_norm_flops_compute)
+
+# poolings
+F.avg_pool1d = wrapFunc(F.avg_pool1d, pool_flops_compute)
+F.avg_pool2d = wrapFunc(F.avg_pool2d, pool_flops_compute)
+F.avg_pool3d = wrapFunc(F.avg_pool3d, pool_flops_compute)
+F.max_pool1d = wrapFunc(F.max_pool1d, pool_flops_compute)
+F.max_pool2d = wrapFunc(F.max_pool2d, pool_flops_compute)
```
Comment (ShadenSmith) [inline deepspeed/profiling/flops_profiler/profiler.py:208]: It would be cool to have a data structure that maps a list of functionals to the compute wrapper, and then at wrap time the engine code just traverses that data structure to do this work.

File: deepspeed/profiling/flops_profiler/profiler.py [hunk 3]
```diff
@@ -0,0 +1,751 @@
...
+
+    def forward(self, x):
+        x = self.feature_extractor(x)
+        x = torch.flatten(x, 1)
+        logits = self.classifier(x)
+        probs = F.softmax(logits, dim=1)
+        return logits, probs
+
+
+if __name__ == "__main__":
```
Comment (ShadenSmith) [inline deepspeed/profiling/flops_profiler/profiler.py:732]: This looks left over from development.

File: deepspeed/runtime/engine.py [hunk 4]
```diff
@@ -38,6 +40,8 @@
 from ..ops.adam import DeepSpeedCPUAdam
 from ..ops.adam import FusedAdam
 
+from deepspeed.profiling.flops_profiler.profiler import add_profile_methods, print_model_profile, print_model_aggregated_profile, flops_to_string, params_to_string
```
Comment (ShadenSmith) [inline deepspeed/runtime/engine.py:43]: Since we manage exports within each subpackage, we can just `from deepspeed.profiling import *` to expose the high-level API I think. That's a bit nicer than having a long list like this.

Since we do have this pattern of exporting a lot of methods, what about wrapping them up in a `FlopsProfiler` class?

**Review by samyam — CHANGES_REQUESTED** [review]
I want to just leave a high level review here based on some our earlier discussion. I think it would be very cool to do the profiling using profiler.start() and profiler.stop() instead of using get_model_profile(...). This is aligned more closely with general training workflows, and additionally the it would not require the user to create an input constructor. Overall, it would be a more natural and seamless way to use the profiler. What do you think @cli99 ?

**Comment by ShadenSmith** [comment]
> I want to just leave a high level review here based on some our earlier discussion. I think it would be very cool to do the profiling using profiler.start() and profiler.stop() instead of using get_model_profile(...). This is aligned more closely with general training workflows, and additionally the it would not require the user to create an input constructor. Overall, it would be a more natural and seamless way to use the profiler. What do you think @cli99 ?

This would also go towards addressing the import side effect we chatted about @cli99. It might be a good opportunity to use [contextlib](https://docs.python.org/3/library/contextlib.html#contextlib.ContextDecorator) ?

File: deepspeed/profiling/flops_profiler/examples/bert.py (hunk 1 above)
Comment (cli99) [reply deepspeed/profiling/flops_profiler/examples/bert.py]: I moved the examples to readme and tests/unit/test_flops_profiler.py

[the remark this answers is not in the document]
Comment (cli99) [reply deepspeed/profiling/flops_profiler/examples/bert.py]: removed the examples. I will add them to dse and readthedocs.

File: tests/small_model_debugging/run.sh [hunk 5]
```diff
@@ -0,0 +1,12 @@
+#! /bin/bash
```
[the remark this answers is not in the document]
Comment (cli99) [reply tests/small_model_debugging/run.sh]: Done

File: deepspeed/profiling/flops_profiler/__init__.py [hunk 6]
```diff
@@ -0,0 +1 @@
+from .profiler import get_model_profile, add_profile_methods, print_model_profile, print_model_aggregated_profile, flops_to_string, duration_to_string, params_to_string
```
[the remark this answers is not in the document]
Comment (cli99) [reply deepspeed/profiling/flops_profiler/__init__.py]: I put all the functions in a FlopsProfiler class which will be the only thing to import

File: deepspeed/profiling/flops_profiler/profiler.py [hunk 7]
```diff
@@ -0,0 +1,751 @@
...
+        flops = funcFlopCompute(*args, **kwds)
+        name = "nn.functional." + func.__name__
+        module_flop_count.append((name, flops))
+        return oldFunc(*args, **kwds)
+
+    return newFunc
+
+
+# FC
+F.linear = wrapFunc(F.linear, linear_flops_compute)
```
[the remark this answers is not in the document]
Comment (cli99) [reply deepspeed/profiling/flops_profiler/profiler.py]: I moved wrapping functionals into the profiler.start() and added unwrapping functionals in the profile.end().

File: deepspeed/profiling/flops_profiler/profiler.py (hunk 2 above)
Comment (cli99) [reply deepspeed/profiling/flops_profiler/profiler.py]: Monkey patching does not work if the target function is referenced as an item in a loop.

File: deepspeed/profiling/flops_profiler/profiler.py (hunk 3 above)
Comment (cli99) [reply deepspeed/profiling/flops_profiler/profiler.py]: Removed

File: deepspeed/runtime/engine.py (hunk 4 above)
Comment (cli99) [reply deepspeed/runtime/engine.py]: Done

File: deepspeed/profiling/flops_profiler/profiler.py [hunk 8]
```diff
@@ -0,0 +1,751 @@
...
+def prod(dims):
+    p = 1
+    for v in dims:
+        p *= v
+    return p
+
+
+def linear_flops_compute(input, weight, bias=None):
+    out_features = weight.shape[0]
+    return torch.numel(input) * out_features
```
[the remark this answers is not in the document]
Comment (cli99) [reply deepspeed/profiling/flops_profiler/profiler.py:269]: the flops profiler outputs MACs and the user can multiply that by 2 or anything. The throughput of a module is computed as TFLOPS where I do MACs*2/time.

File: deepspeed/profiling/flops_profiler/profiler.py [hunk 9]
```diff
@@ -0,0 +1,751 @@
...
+
+
+def prod(dims):
+    p = 1
+    for v in dims:
+        p *= v
+    return p
+
+
+def linear_flops_compute(input, weight, bias=None):
```
[the remark this answers is not in the document]
Comment (cli99) [reply deepspeed/profiling/flops_profiler/profiler.py]: what would be the flops for computing bias? I thought it's zero.

[the remark this answers is not in the document]
Comment (RezaYazdaniAminabadi) [reply deepspeed/profiling/flops_profiler/profiler.py]: You can consider the dimension of bias (like out_features), which is the number of addition. But, this will be very small probably, so that's okay if not considering that.

File: deepspeed/profiling/flops_profiler/profiler.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (RezaYazdaniAminabadi) [reply deepspeed/profiling/flops_profiler/profiler.py:308]: I see, but I was thinking that this is not always MAC. I am not sure if we want to take into account the point-wise operations into the profiler like bias-add or normalize.

**Review by samyam — APPROVED** [review]

File: deepspeed/profiling/flops_profiler/README.md [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+profile_start_setp = 5
+profile_end_step = 10
+assert (profile_end_step > profile_start_step), "should end profiling after start profiling"
+print_profile = True
+pring_aggregated_profile = True
+
+for step, batch in enumerate(data_loader):
+  # start profiling at training step "profile_step"
+  if step == profile_start_step:
+    model.start_profile()
```
[the remark this answers is not in the document]
Comment (cli99) [reply deepspeed/profiling/flops_profiler/README.md]: The readme shown here is out of date. The latest code does this. Happy to chat more about it.

**Comment by jeffra** [comment]
Closing this one after rebasing/squashing, moved to #664
